core/system/src/export_table.c

```diff
@@ -14,7 +14,105 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include <uvisor.h>
 #include "api/inc/export_table_exports.h"
+#include "api/inc/vmpu_exports.h"
+#include "context.h"
+#include "halt.h"
+
+/* By default a maximum of 16 threads are allowed. This can only be overridden
+ * by the porting engineer for the current platform. */
+#ifndef UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT
+#define UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT ((uint32_t) 16)
+#endif
+
+/* Per thread we store the pointer to the allocator and the process id that
+ * this thread belongs to. */
+typedef struct {
+    void * allocator;
+    int process_id;
+} UvisorThreadContext;
+
+static UvisorThreadContext thread[UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT];
+
+
+static int thread_ctx_valid(UvisorThreadContext * context)
+{
+    /* Check if context pointer points into the array. */
+    if ((uint32_t) context < (uint32_t) &thread ||
+        ((uint32_t) &thread + sizeof(thread)) <= (uint32_t) context) {
+        return 0;
+    }
+    /* Check if the context is aligned exactly to a context. */
+    return (((uint32_t) context - (uint32_t) thread) % sizeof(UvisorThreadContext)) == 0;
+}
+
+static void * thread_create(int id, void * c)
+{
+    (void) id;
+    UvisorThreadContext * context = c;
+    const UvisorBoxIndex * const index =
+            (UvisorBoxIndex * const) *(__uvisor_config.uvisor_box_context);
+    /* Search for a free slot in the tasks meta data. */
+    uint32_t ii = 0;
+    for (; ii < UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT; ii++) {
+        if (thread[ii].allocator == NULL)
+            break;
+    }
+    if (ii < UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT) {
+        /* Remember the process id for this thread. */
+        thread[ii].process_id = g_active_box;
+        /* Fall back to the process heap if ctx is NULL. */
+        thread[ii].allocator = context ? context : index->box_heap;
```

Comment:
@niklas-arm @AlessandroA that looks insecure: using the user-controllable box context pointer in global memory instead of using the one from the internal table.